### ISSUE TITLE
KML/GeoJSON polygon onFeatureClick() not called

### ISSUE THREAD

**Issue by tenSunFree** [googlemaps/android-maps-utils#558]
關鍵在於 Renderer 的 addedPolygon.setClickable(polygonOptions.isClickable());

polygonOptions.isClickable() 總是得到false

所以如果想讓 KmlDemoActivity的onFeatureClick 可以被調用

必須修正成 addedPolygon.setClickable(true);

**Comment by tenSunFree** [comment]
PolygonOptions在創建的時候 zzcu的值預設為false

也就是說 PolygonOptions.isClickable() 預設為false

也就導致 Renderer的 addedPolygon.setClickable(false) 

進而導致 onFeatureClick 不會被調用

![2019-10-08_020908](https://user-images.githubusercontent.com/23090085/66337314-80068d00-e971-11e9-8124-d9ff1704c925.jpg)

**Comment by barbeau** [comment]
I looked at this with the current master branch at https://github.com/googlemaps/android-maps-utils/commit/80d9497ed10b9ec92e16376008b1727ff6695eea (0.6.2), and here's my understanding of what's happening (partially thanks to Google Translate). @tenSunFree Please let me know if I got any of this wrong :).

#### Summary

In `KmlDemoActivity`, if you load the KML file from a local resource and then tap on it, nothing happens (`KmlLayer.OnFeatureClickListener`'s `onFeatureClick()` is never called).

#### Steps to repro

1. In `KmlDemoActivity.startDemo()`, comment out `retrieveFileFromUrl();` and uncomment `retrieveFileFromResource();` to load a KML file as a map layer from bundled resources instead of a remote URL.
1. Start the demo app and tap on "KML LAYER OVERLAY" button at the bottom
1. Tap on one of the Google building map polygons (e.g., the blue polygon)

#### What I expect to see

The click listener should fire and I should see a Toast that says `Feature clicked: #`, where `#` is the feature ID.

#### What I actually see

Tapping on the blue polygon has no effect - the listener doesn't fire (see below screenshot).

![image](https://user-images.githubusercontent.com/928045/67899123-06109080-fb38-11e9-9ce5-4d939baaa633.png)

I also noticed that the remote URL used for `retrieveFileFromUrl();` is broken, but that's unrelated to this issue, so I'll open a separate issue for this.

**Comment by barbeau** [comment]
I ended up digging further into this while reviewing PR https://github.com/googlemaps/android-maps-utils/pull/380 (PR https://github.com/googlemaps/android-maps-utils/pull/380 does not appear to fix this issue).

The root problem here seems to be that the Maps API PolygonOptions.clickable attribute defaults to false:
https://developers.google.com/android/reference/com/google/android/gms/maps/model/PolygonOptions.html#clickable(boolean)

...as does PolylineOptions.clickable:
https://developers.google.com/android/reference/com/google/android/gms/maps/model/PolylineOptions.html#clickable(boolean)

In `Render.addLineStringToMap()` there is a `addedPolyline.setClickable(true);`, which enables the click listener to fire.

However, there isn't an equivalent in `Renderer.addPolygonToMap()` - instead it just copies the existing `PolygonOptions` clickable attribute with `addedPolygon.setClickable(polygonOptions.isClickable());`.

This affects GeoJSON polygons too, but the difference being that in app code you can set clickable to true for GeoJSON polygons by setting a GeoJSON polygon style:

~~~
GeoJsonPolygonStyle polygonStyle = new GeoJsonPolygonStyle();
polygonStyle.setClickable(true);
~~~

...which then gets propagated to `Renderer.addPolygonToMap()`.

Unless I'm missing something, there doesn't seem to be an equivalent way to set clickable to true from app code via `KmlLayer`, `KmlPlacemark`, or `KmlStyle`, and hence there isn't a way to make the KML polygon clickable.

Possible solution seems to be adding a boolean clickable attribute to `KmlLayer` constructors and passing that through to the `KmlStyle` so it gets set in the PolygonOptions at some point.

**Comment by barbeau** [comment]
I'm also wondering if the hard-coding of `addedPolyline.setClickable(true);` in `Render.addLineStringToMap()` is a mistake - seems like this would override app code like:

~~~
// Create a new polygon style
GeoJsonLineStringStyle lineStyle = new GeoJsonLineStringStyle();
lineStyle.setClickable(false);
~~~

**Comment by barbeau** [comment]
I was curious to see if there were any interactions with the KML spec itself here - the closest attribute I can find is BalloonStyle `<displayMode>`, [which says](https://developers.google.com/kml/documentation/kmlreference#elements-specific-to-balloonstyle):

>If <displayMode> is default, Google Earth uses the information supplied in <text> to create a balloon . If <displayMode> is hide, Google Earth does not display the balloon. In Google Earth, clicking the List View icon for a Placemark whose balloon's <displayMode> is hide causes Google Earth to fly to the Placemark.

Regardless of how this should translate to this library, the documentation for this library clearly says that `<displayMode>` isn't supported, so we don't need to worry about supporting that in fixing this issue:
https://developers.google.com/maps/documentation/android-sdk/utility/kml#supported

**Comment by barbeau** [comment]
I wanted to see where exactly this issue regressed.

Looking back at old library versions:
* v0.4.4 - [KmlDemoActivity](https://github.com/googlemaps/android-maps-utils/blob/0.4.4/demo/src/com/google/maps/android/utils/demo/KmlDemoActivity.java#L98) doesn't include the line `kmlLayer.setOnFeatureClickListener()`, and `KmlLayer` doesn't actually include that method.
* v0.5.0 - First release to include the line `kmlLayer.setOnFeatureClickListener()` in [KmlDemoActivity](https://github.com/googlemaps/android-maps-utils/blob/0.5.0/demo/src/com/google/maps/android/utils/demo/KmlDemoActivity.java#L103). But tapping on polygon is still broken.

Using Git blame, it looks like the KML click listening in KmlDemoActivity was added as part of the GeoJSON/KML integration in https://github.com/googlemaps/android-maps-utils/pull/351. https://github.com/googlemaps/android-maps-utils/pull/351#issuecomment-275014634 says:

>The changes here refactors this so that the KML and GeoJSON implementations share a common base abstraction. This should make implementing features and fixes for both a lot more easier in the future. For example, last year I implemented click handling for GeoJSON, but not KML, given the previous design, but that's now been addressed as part of this new work.

I tested with the specific PR merge commit (https://github.com/googlemaps/android-maps-utils/commit/a7a8fcc5660c569c59a9fb1e338326b40bbd2be3), but same result as v0.5.0 - tapping on the KML polygon has no effect.

So, it appears that KML polygon click listening never actually worked as intended, so we're free to come up with our own solution here.

**Comment by barbeau** [comment]
Actually, looking more carefully at the original KML/GeoJSON integration PR, it did include [the line to hard-code clickable to true](https://github.com/googlemaps/android-maps-utils/pull/351/files#diff-dea26f53a40d31f061e3537537389db6R683) for polygons.

~~~
    /**
     * Adds a DataPolygon to the map as a Polygon
     *
     * @param polygonOptions
     * @param polygon      contains coordinates for the Polygon
     * @return Polygon object created from given DataPolygon
     */
    protected Polygon addPolygonToMap(PolygonOptions polygonOptions, DataPolygon polygon) {
        // First array of coordinates are the outline
        polygonOptions.addAll(polygon.getOuterBoundaryCoordinates());
        // Following arrays are holes
        List<List<LatLng>> innerBoundaries = polygon.getInnerBoundaryCoordinates();
        for (List<LatLng> innerBoundary : innerBoundaries) {
            polygonOptions.addHole(innerBoundary);
        }
        Polygon addedPolygon = mMap.addPolygon(polygonOptions);
        addedPolygon.setClickable(true);
        addedPolygon.setClickable(polygonOptions.isClickable());
        return addedPolygon;
    }
~~~

...but then [this PR](https://github.com/googlemaps/android-maps-utils/pull/454) allowed setting the clickable attribute via the style, changing the above code to:

~~~
addedPolygon.setClickable(polygonOptions.isClickable());
~~~

...which broke it, because the PolygonOptions defaults clickable to false. So the issue was re-introduced prior to the 0.6.0 release.

So, in summary, we effectively have the same issue for both KML/GeoJSON polylines and polygons right now:
* We don't currently support both a developer-defined clickable attribute as well as a default clickable value of true.

...but we're handling each case differently:
* For lines - we're currently setting default clickable attribute to true and ignoring developer-defined clickable attribute
* For polygons - we're currently supporting developer-defined clickable attributes, but that results in a default of not clickable. And for KmlLayer there currently isn't a way to override the default.
